Bug: The engine crashes during write operation

### What happened?

After updating on k8s cluster, the engine stops working. When a client application tries to write a record, it prints an error `[500] Bad block` and crushes.

### Version

0.8.0-dev

### OS

Linux

### Snippet to repeat the bug

_No response_

### Logs

```Shell
ERROR: ld.so: object '/usr/local/lib/AppProtection/libAppProtection.so' from /etc/ld.so.preload cannot be preloaded (failed to map segment from shared object): ignored.
2022-08-01 15:32:17.460 ( 7168)  [INFO] -- main.cc:41 Reduct Storage 0.8.0 
2022-08-01 15:32:17.460 ( 7168)  [INFO] -- main.cc:55 Configuration: 
 	RS_LOG_LEVEL = INFO 
	RS_HOST = 0.0.0.0 (default)
	RS_PORT = 8383 (default)
	RS_API_BASE_PATH = / (default)
	RS_DATA_PATH = /data (default)
	RS_API_TOKEN = ********** (masked)
 
2022-08-01 15:32:17.639 ( 7168)  [INFO] -- storage.cc:39 Load 1 buckets 
2022-08-01 15:32:17.639 ( 7168)  [INFO] -- api_server.cc:143 Run HTTP server on http://0.0.0.0:8383/ 
2022-08-01 15:32:31.737 ( 7168) [ERROR] -- common.h:145 POST /b/data/acc-4: [500] Bad block 

Process finished with exit code 0
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,15 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-## Fixed:
+### Fixed:
 
 - Infinite loop in Bucket::KeepQuota, [PR-146](https://github.com/reduct-storage/reduct-storage/pull/146)
 - waiting data from HTTP client if it aborts
   connection, [PR-151](https://github.com/reduct-storage/reduct-storage/pull/151)
 
-## Changed:
+### Changed:
 
-- Duplication of timestamps is not allowed, [PR147](https://github.com/reduct-storage/reduct-storage/pull/147)
+- Duplication of timestamps is not allowed, [PR-147](https://github.com/reduct-storage/reduct-storage/pull/147)
+
+### Fixed:
+
+- writing record when current block is broken, [PR-15-](https://github.com/reduct-storage/reduct-storage/pull/153)
 
 ## [0.7.1] - 2022-07-30
 

--- a/src/reduct/core/time.h
+++ b/src/reduct/core/time.h
@@ -7,5 +7,10 @@
 
 namespace reduct::core {
 using Time = std::chrono::system_clock::time_point;
+
+inline auto ToMicroseconds(Time tp) {
+  return std::chrono::duration_cast<std::chrono::microseconds>(tp.time_since_epoch()).count();
+}
+
 }
 #endif  // REDUCT_STORAGE_TIME_H

--- a/src/reduct/proto/storage/entry.proto
+++ b/src/reduct/proto/storage/entry.proto
@@ -10,6 +10,7 @@ message Record {
     kStarted = 0;     // has at least one chunk written
     kFinished = 1;    // finished without errors
     kErrored = 2;     // finished with error
+    kInvalid = 3;     // something wierd happened
   }
 
   message MetaEntry {
@@ -31,4 +32,5 @@ message Block {
   google.protobuf.Timestamp latest_record_time = 2;    // the timestamp of the latest record
   uint64 size = 3;                                // size of block in bytes (with protobuf overhead)
   repeated Record records = 4;                    // stored records
+  bool invalid = 5;                               // mark block as invalid if some IO happened
 }

--- a/src/reduct/storage/block_manager.cc
+++ b/src/reduct/storage/block_manager.cc
@@ -151,6 +151,10 @@ class BlockManager : public IBlockManager {
           if (auto err = SaveBlock(blk)) {
             LOG_ERROR("{}", err.ToString());
           }
+
+          if (state == proto::Record::kInvalid) {
+            blk->set_invalid(true);
+          }
         });
 
     auto& writers = RemoveDeadWriters(block);

--- a/src/reduct/storage/block_manager.cc
+++ b/src/reduct/storage/block_manager.cc
@@ -105,8 +105,6 @@ class BlockManager : public IBlockManager {
   }
 
   Error RemoveBlock(const BlockSPtr& block) override {
-    std::error_code ec;
-
     const auto& readers = RemoveDeadReaders(block);
     if (!readers.empty()) {
       return {.code = 500, .message = "Block has active readers"};
@@ -117,17 +115,27 @@ class BlockManager : public IBlockManager {
       return {.code = 500, .message = "Block has active writers"};
     }
 
-    fs::remove(BlockPath(parent_, *block), ec);
+    std::error_code ec;
+    auto path = BlockPath(parent_, *block);
+    auto make_error = [&ec, &path] {
+      return Error{.code = 500, .message = fmt::format("Failed to remove block {}: {}", path.string(), ec.message())};
+    };
+
+    // remove block with date
+    fs::remove(path, ec);
+    Error err;
     if (ec) {
-      return {.code = 500, .message = ec.message()};
+      err = make_error();
     }
 
-    fs::remove(BlockPath(parent_, *block, kMetaExt), ec);
+    // remove descriptor
+    path = BlockPath(parent_, *block, kMetaExt);
+    fs::remove(path, ec);
     if (ec) {
-      return {.code = 500, .message = ec.message()};
+      err = make_error();
     }
 
-    return Error::kOk;
+    return err;
   }
 
   core::Result<async::IAsyncReader::SPtr> BeginRead(const BlockSPtr& block, AsyncReaderParameters params) override {
@@ -147,13 +155,14 @@ class BlockManager : public IBlockManager {
             LOG_ERROR("{}", load_err.ToString());
             return;
           }
-          blk->mutable_records(index)->set_state(state);
-          if (auto err = SaveBlock(blk)) {
-            LOG_ERROR("{}", err.ToString());
-          }
 
+          blk->mutable_records(index)->set_state(state);
           if (state == proto::Record::kInvalid) {
             blk->set_invalid(true);
+          }
+
+          if (auto err = SaveBlock(blk)) {
+            LOG_ERROR("{}", err.ToString());
           }
         });
 

--- a/src/reduct/storage/entry.cc
+++ b/src/reduct/storage/entry.cc
@@ -308,6 +308,7 @@ class Entry : public IEntry {
     }
     auto stop_ts = FromTimePoint(query_info.stop);
 
+    // Find start block
     auto start_block = block_set_.upper_bound(start_ts);
     if (start_block == block_set_.end()) {
       start_block = std::prev(start_block);
@@ -315,10 +316,27 @@ class Entry : public IEntry {
       start_block = std::prev(start_block);
     }
 
-    const auto [block, err] = block_manager_->LoadBlock(*start_block);
+    auto [block, err] = block_manager_->LoadBlock(*start_block);
     if (err) {
       queries_.erase(query_id);
       return {{}, err};
+    }
+
+    // Check if it is valid
+    if (block->invalid()) {
+      start_block = std::next(start_block);
+      if (start_block != std::end(block_set_) && *start_block < stop_ts) {
+          const auto [next_block, next_err] = block_manager_->LoadBlock(*start_block);
+          if (next_err) {
+            queries_.erase(query_id);
+            return {{}, next_err};
+          }
+
+          block = next_block;
+      } else {
+        queries_.erase(query_id);
+        return {{}, {.code = 202, .message = "No Content"}};
+      }
     }
 
     std::vector<int> records;

--- a/src/reduct/storage/entry.cc
+++ b/src/reduct/storage/entry.cc
@@ -130,11 +130,10 @@ class Entry : public IEntry {
     auto has_no_space = block->size() + content_size > options_.max_block_size;
     auto too_many_records = block->records_size() + 1 > options_.max_block_records;
 
-    if (type == RecordType::kLatest && (has_no_space || too_many_records)) {
+    if (type == RecordType::kLatest && (has_no_space || too_many_records || block->invalid())) {
       LOG_DEBUG("Create a new block");
       if (auto err = block_manager_->FinishBlock(block)) {
-        LOG_ERROR("Failed to finish the current block");
-        return {{}, err};
+        LOG_WARNING("Failed to finish the current block: {}", err.ToString());
       }
 
       auto ret = start_new_block(proto_ts, content_size);

--- a/src/reduct/storage/io/async_writer.cc
+++ b/src/reduct/storage/io/async_writer.cc
@@ -33,7 +33,7 @@ class AsyncWriter : public async::IAsyncWriter {
   Error Write(std::string_view chunk, bool last) noexcept override {
     const auto& record = parameters_.record_index;
     if (!file_) {
-      update_record_(record, proto::Record::kErrored);
+      update_record_(record, proto::Record::kInvalid);
       return {.code = 500, .message = "Bad block"};
     }
 
@@ -44,7 +44,7 @@ class AsyncWriter : public async::IAsyncWriter {
     }
 
     if (!file_.write(chunk.data(), chunk.size())) {
-      update_record_(record, proto::Record::kErrored);
+      update_record_(record, proto::Record::kInvalid);
       return {.code = 500, .message = "Failed to write a chunk into a block"};
     }
 

--- a/unit_tests/reduct/storage/entry_query_test.cc
+++ b/unit_tests/reduct/storage/entry_query_test.cc
@@ -13,6 +13,7 @@ using reduct::WriteOne;
 using reduct::async::IAsyncReader;
 using reduct::core::Error;
 using reduct::core::Time;
+using reduct::core::ToMicroseconds;
 using reduct::storage::IEntry;
 using reduct::storage::query::IQuery;
 
@@ -22,6 +23,7 @@ using std::chrono::seconds;
 namespace fs = std::filesystem;
 
 const auto kName = "entry_1";
+const IQuery::Options kDefaultOptions = {.ttl = seconds(1)};
 
 static auto MakeDefaultOptions() {
   return IEntry::Options{
@@ -35,8 +37,6 @@ static const auto kTimestamp = Time() + std::chrono::microseconds(10'100'200);  
 TEST_CASE("storage::Entry should query records", "[entry][query]") {
   auto entry = IEntry::Build(kName, BuildTmpDirectory(), MakeDefaultOptions());
   REQUIRE(entry);
-
-  const IQuery::Options kDefaultOptions = {.ttl = seconds(1)};
 
   SECTION("empty entry") {
     auto [start, stop] =
@@ -197,8 +197,6 @@ TEST_CASE("storage::Entry should have TTL", "[entry][query]") {
   auto entry = IEntry::Build(kName, BuildTmpDirectory(), MakeDefaultOptions());
   REQUIRE(entry);
 
-  const IQuery::Options kDefaultOptions = {.ttl = seconds(1)};
-
   REQUIRE(WriteOne(*entry, "blob", kTimestamp) == Error::kOk);
   REQUIRE(WriteOne(*entry, "blob", kTimestamp + seconds(1)) == Error::kOk);
   REQUIRE(WriteOne(*entry, "blob", kTimestamp + seconds(2)) == Error::kOk);
@@ -228,4 +226,24 @@ TEST_CASE("storage::Entry should have TTL", "[entry][query]") {
                        .message = fmt::format("Query id={} doesn't exist. It expired or was finished", id),
                    });
   }
+}
+
+TEST_CASE("storage::Entry should handle invalid blocks", "[entry][query]") {
+  auto const path = BuildTmpDirectory();
+  auto entry = IEntry::Build(kName, path, MakeDefaultOptions());
+
+  std::string data = "xxxx";
+  REQUIRE(WriteOne(*entry, data, kTimestamp) == Error::kOk);
+  // remove block with data to make write operation invalid
+  REQUIRE(fs::remove(path / kName / fmt::format("{}.blk", ToMicroseconds(kTimestamp))));
+  REQUIRE(WriteOne(*entry, data, kTimestamp + seconds(1)) == Error{.code = 500, .message = "Bad block"});
+
+  REQUIRE(WriteOne(*entry, data, kTimestamp + seconds(2)) == Error::kOk);
+
+  auto [id, _] = entry->Query(kTimestamp, kTimestamp + seconds(3), kDefaultOptions);
+
+  auto [record, err] = entry->Next(id);
+  REQUIRE(err == Error::kOk);
+  REQUIRE(record.reader->timestamp() == kTimestamp + seconds(2));
+  REQUIRE(record.last);
 }

--- a/unit_tests/reduct/storage/entry_test.cc
+++ b/unit_tests/reduct/storage/entry_test.cc
@@ -13,6 +13,7 @@ using reduct::ReadOne;
 using reduct::WriteOne;
 using reduct::core::Error;
 using reduct::core::Time;
+using reduct::core::ToMicroseconds;
 using reduct::storage::IEntry;
 
 using google::protobuf::util::TimeUtil;
@@ -31,15 +32,9 @@ static auto MakeDefaultOptions() {
 
 static const auto kTimestamp = Time() + std::chrono::microseconds(10'100'200);  // to check um precision
 
-auto ToMicroseconds(Time tp) {
-  return std::chrono::duration_cast<std::chrono::microseconds>(tp.time_since_epoch()).count();
-}
 
 auto GetBlockSize(std::string_view name, fs::path path, IEntry::Options options, Time begin_ts) {
-  return fs::file_size(
-      path / name /
-      fmt::format("{}.blk",
-                  std::chrono::duration_cast<std::chrono::microseconds>(begin_ts.time_since_epoch()).count()));
+  return fs::file_size(path / name / fmt::format("{}.blk", ToMicroseconds(begin_ts)));
 }
 
 TEST_CASE("storage::Entry should record data to a block", "[entry]") {
@@ -165,6 +160,22 @@ TEST_CASE("storage::Entry should create block with size of record it  is bigger 
   REQUIRE(WriteOne(*entry, big_data, kTimestamp) == Error::kOk);
 
   REQUIRE(GetBlockSize(kName, path, options, kTimestamp) == big_data.size());
+}
+
+TEST_CASE("storage::Entry should create block if previous invalid", "[entry][block]") {
+  const auto options = MakeDefaultOptions();
+  const auto path = BuildTmpDirectory();
+  auto entry = IEntry::Build(kName, path, options);
+  REQUIRE(entry);
+
+  std::string data = "xxxx";
+  REQUIRE(WriteOne(*entry, data, kTimestamp) == Error::kOk);
+  // remove block with data to make write operation invalid
+  REQUIRE(fs::remove(path / kName / fmt::format("{}.blk", ToMicroseconds(kTimestamp))));
+  REQUIRE(WriteOne(*entry, data, kTimestamp + seconds(1)) == Error{.code = 500, .message = "Bad block"});
+
+  REQUIRE(WriteOne(*entry, data, kTimestamp + seconds(2)) == Error::kOk);
+  REQUIRE(GetBlockSize(kName, path, options, kTimestamp + seconds(2)) == 100);
 }
 
 TEST_CASE("storage::Entry should write data for random kTimestamp", "[entry]") {

